File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.java
import java.util.stream.*;
import java.util.*;

class Solution {
    public int maxSheep = 0;
    public int[][] child;
    public int[] info;
    
    public int solution(int[] info, int[][] edges) {
        child = new int[info.length][2];
        this.info = info;
        
        for (int[] ab : edges) {
            int idx = 1;
            if (child[ab[0]][0] == 0) {
                idx = 0;
            }
            child[ab[0]][idx] = ab[1];
        }
        
        Set<Integer> set = new HashSet<Integer>();
        set.add(0);
        dfs(1, 0, set);
        return maxSheep;
    }
    
    public void dfs(int sheepCnt, int wolfCnt, Set<Integer> nodes){
        if (wolfCnt >= sheepCnt){
            return;
        }
        
        maxSheep = Math.max(sheepCnt, maxSheep);
        
        for(int node : nodes){
            for(int nextNode : child[node]){
                if (nextNode == -1) continue;
                if (nodes.contains(nextNode)) continue;
                
                Set<Integer> set = new HashSet<>(nodes);
                set.add(nextNode);
                
                if(info[nextNode] == 1){
                    dfs(sheepCnt, wolfCnt + 1, set);
                } else {
                    dfs(sheepCnt + 1, wolfCnt, set);
                }
            }
        }
    }
}

File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.py
def solution(info, edges):
    from collections import defaultdict, deque

    def bfs(info, graph):
        max_sheep = 0
        queue = deque([(0, 1, 0, [0])])

        while queue:
            node, sheep, wolf, route = queue.popleft()
            max_sheep = max(max_sheep, sheep)

            for next_node in route:
                for child in graph[next_node]:
                    if child not in route:
                        next_sheep = sheep + (info[child] == 0)
                        next_wolf = wolf + (info[child] == 1)

                        if next_wolf < next_sheep:
                            queue.append((child, next_sheep, next_wolf, route + [child]))

        return max_sheep

    graph = defaultdict(list)
    for parent, child in edges:
        graph[parent].append(child)

    return bfs(info, graph)

